### ISSUE TITLE
Is the view.show() method being abused?

### ISSUE THREAD

**Issue by theScottyJam** [robertbraeutigam/tictactoe#4]
In `HumanPlayer.java`, you have this line:

```java
view.show(cells -> ui.askForMove().selectFrom(cells).mark());
```

Isn't this sort of an abuse of what `view.show()` is meant to do? To me, a "show" function with this sort of signature is supposed to give its data to a medium that then shows it somehow (whether that's to stdout, a log file, etc). Here, we're not trying to show the cells to any medium, instead, we're just trying to do some business logic against the cells. In other words, even though this method is called "show", we're basically using it as a warped "getCells()" method. In this particular scenario, calling the method "getCells()" would better explain what's going on.

```java
view.getCells(cells -> ui.askForMove().selectFrom(cells).mark());
```

We're just, for some funny reason, passing the cells to a provided lambda instead of just returning them to the caller.

Does this make sense? I've noticed you have an article on avoiding getters, and I haven't read through it yet (I well), but I have seen other people's articles on that subject. I assume this code ended up this way because of an attempt to avoid getters, and yet, I think you've still got a getter in here, disguised under a different "action" name, and providing its value to a lambda instead of returning it.

(Maybe, you wouldn't consider it a getter, even if it was called getCells(), because it's technically constructing a brand new 2d cell array everytime you call it, so it's not getting internal data, it's returning something new. But, from what I can tell, that's technically just an implementation detail. Nothing about this function makes it apparent that it's always providing a new value, in fact, you could trivially move this 2d cell array construction logic to SquareView's constructor and give the same 2d cell array to everyone who calls show(), and no one would be none the wiser).

**Comment by robertbraeutigam** [comment]
I tend to agree with your comment. It is a getter, returning mostly data.

As discussed in the other issue, `View` is a boundary for external code. We need to supply some data to the external code for use-cases we don't yet know about.

I might change it to `getCells()` as you noted. Or maybe I just redesign the whole thing without the need to be externally extensible. I think the latter would make it clearer how OO designs would normally look like. Or maybe you want to take a crack at it as an exercise?

**Comment by theScottyJam** [comment]
> Or maybe I just redesign the whole thing without the need to be externally extensible. I think the latter would make it clearer how OO designs would normally look like.

Yeah, I think so too, as it's a bit difficult to see which places you're intentionally trying to plan for flexibility, and what's unintentional.

> Or maybe you want to take a crack at it as an exercise?

That would be fun :). I'll give it a shot. You'll probably need to do a fair amount of peer-reviewing on it though, as I'm still trying to grapple with this coding style, and when you should and shouldn't apply different principles.

**Comment by theScottyJam** [comment]
hmm... though I'm not sure how fruitful this would be either.

If we're not planning for extensibility, the project might come down to something that's pretty simple and not very OOP-y.

If we think about it, there's not much state involved in the project:
* The game board
* Who's turn it is

In the current project, you have a single class that holds the entire state of the game board (SqureBoard.java), and the player's turn resides in Game.java. All of the rest of the logic simply operates on that state. If we don't need this to be extensible, then if I understand correctly, we're going to end up moving all of this logic that's coupled with the game board into SquareBoard.java, and since there's really not much logic related to who's turn it is (at least, there doesn't need to be), Game.java will remain fairly small. And, we'll end up with a two-class project (if you exclude the classes from main/). That doesn't do a very good job of showcasing OOP.

I guess you could still try and create HumanPlayer and RandomPlayer classes, but all of their business logic will still need to sit within Game.java, because their business logic is coupled to the game board's state (everything except the logic to prompt the user for a position, that's coupled to the player's turn because the prompt message shows who's turn it is, so that would reside in Game.java)

So, perhaps a better showcase would be a program that had a bit more state in it? So that there was a bit more stuff to encapsulate and separate from each other? Like, a TODO list app? Or, do you feel like I'm misjudging how the tic-tac-toe app would be best organized?

**Comment by theScottyJam** [comment]
On second thought, it's trivial to pump the tic-tac-toe game with a little more state if needed. For example, I could have a small selection screen to let you choose if it's one player against a computer, or two player.

**Comment by theScottyJam** [comment]
Alright, I've finally gotten around to doing a rewrite which you can find [here](https://github.com/robertbraeutigam/tictactoe/pull/5). I decided to do it almost completely from scratch, then see what I ended up with. I did my best to follow what I perceive to be the type of OOP principles you're promoting, I especially focused on trying to follow the law of Demeter as strictly as reasonably possible. I also followed other principles, like show-don't-tell, dependency inversion, etc.

I ended up with a solution that's likely over-engineered - it certainly feels bloated compared to the amount of work it should take to make a game like this. I'm hoping for some feedback on how you would have coded this up differently, or how you would simplify things if you were doing this.

I also added a suite of unit tests, in order to test the test-ability of this code. You don't have to pay as much attention to these, and I might just toss them in future drafts if maintaining them gets in the way too often.

Here are a handful of specific questions/concerns I had as I worked on this project.
* Do you apply the law of Demeter apply to built-in data types? For example, I chose to use an optional data type, despite the fact that it forces me to break the law everywhere (I often have to do `optional.orElseThrow().myMethod()`). I justified the use of the Optional data type because I figured it's better than using `null` all over the place. Plus, in other languages the optional data type would be a simple algebraic data type, and unwrapping those would require the use of pattern-matching syntax instead of sending messages to the optional instance, meaning I would be allowed to use optionals without breaking the LoD in other languages. Likewise, I noticed that lists force me to break the law - I would have to inherit from an existing List implementation if I wanted to obey the law here, and I didn't particularly want to do that in this project.
* Strict adherence to the LoD makes it feel like the overall structure of my app is very fragile, and causes the project structure to depend on things it really shouldn't depend on. Some examples:
  * Whether or not a library (including the standard library) chooses to implement a method as a static method or as instance method would influence how I structure my app. For example, say I'm using third-party user-management library that provides `User`, `Group`, and `Sync` classes, then `Sync.resync(myGroup.getUserById(id))` is legal, but if the library decides to deprecate the `Sync` class in favor on instance methods, then I'd have to refactor my project and possibly break out new classes, because the recommended alternative, `myGroup.getUserById(id).resync()`, is illegal. The fact that my app structure depends on the structure of the third-party library feels odd to me.
  * My code structure is also greatly influenced by the syntactic features my language of choice provides. Java, for example, doesn't provide a way to return multiple values (unlike C# which has `out` parameters). Because of this, and because of your dislike towards records, you'd have to create a new class to package these values together before returning them, at which point the LoD comes into play and required logic to be refactored and moved into the new class, all because Java doesn't provide a way to do multiple-value returns.
* The TurnTracker class can't just return an instance of the current Player, so I found myself recreating a number of pass-through functions that would simply send call a function on the current Player with the same arguments provided. This feels like a hack of a design, that's just working around the limitations of the LoD.
* This whole OOP paradigm works fine when you are able to subdivide an algorithm into smaller parts, where each part only interacts with a certain type of data, then you place each piece of the algorithm on the different classes they're associated with. But, it all seems to fall apart as soon as the algorithm needs to operate on multiple types of data at the same time. For example, the action of occupying a tic-tac-toe cell with an X or an O requires ones class to give up some internal details to another class. Either the Position class has to give up it's internal x and y coordinates to the Board class, or the Board class must give up it's internal 2d array to the Position class. It just feels like I'm jumping through hoops, trying to encapsulate this data as much as possible, and yet my hand is still forced to give up internal details to other classes whenever a particular algorithm requires it. Granted, yes, I can tell by looking at the class who's I'm giving my internal data to, but in order to get that sense of safety you talked about [here](https://javadevguy.wordpress.com/2019/09/20/java-records-making-bad-designs-more-convenient/), where it's easy to know who uses the internal data, I'd have to make sure I also jump into the Board class to see if it passes the data along to anyone else, or if it simply uses it and throws it away. We're lucky in this scenario that the Board simply chooses to use it and be done, but it might not always be this way.

Anyways, those are some of the main concerns that surfaced as I tried to work on this project. But, let me know what you think, and if you feel like I was doing anything wrong here, or if I was following any particular guideline in an over-zealous way.
